### PR TITLE
Implemented new typesystem.

### PR DESCRIPTION
**Pull request by azreika**
Implemented new and improved typesystem, based on a newly devised type-lattice concept.

Helper documentation to come.

Combined work with @olligobber.

### CODE REVIEW

File: src/AnalysisType.h [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+ * A collection of analysis types that may appear in a type lattice.
+ *
+ ***********************************************************************/
+
+#pragma once
+
+#include "AstType.h"
+#include "Util.h"
+#include <cassert>
+#include <sstream>
```
Comment (mmcgr) [inline src/AnalysisType.h]: should this be ostream?
Also need vector, set, and memory, and maybe some more Ast stuff?

Running iwyu would be a good idea. Something like
```
for i in *.cpp;do iwyu --Xiwyu --transitive_includes_only --std=c++11 -I. -I.. -I ~/souffle/include $i 2>>iwyuT.log;done
```

**Comment by mmcgr** [comment]
Running clang-tidy would also be good. I use a script like this (not sure if it's still necessary)
```
#!/bin/sh
#MPI="-std=c++11 -DUSE_PROVENANCE -DUSE_MPI -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -I/usr/lib/x86_64-linux-gnu/openmpi/include -pthread  -DUSE_LIBZ  -DUSE_SQLITE  -fopenmp "

if [ -z "$1" ]
then
    FILES="*.cpp"
else
    FILES="$1"
fi
echo "checking $FILES"

for FILE in $FILES
do
clang-tidy $FILE -fix -header-filter="[^j]*[a-z]\.h$" -extra-arg=-std=c++11 \
    -- -I/usr/include/c++/7/ \
    -I/usr/include/x86_64-linux-gnu/c++/7/ -I.. -I. -I../include \
    $MPI
done
```

File: src/AnalysisType.h (hunk 1 above)
Comment (mmcgr) [reply src/AnalysisType.h]: Incidentally, don't assume iwyu is correct. Only apply what makes sense.

File: src/ComponentModel.cpp [hunk 2]
```diff
@@ -310,8 +310,21 @@ void collectContent(const AstComponent& component, const TypeBinding& binding,
...
+                    const_cast<AstTypeCast&>(cast).setType(newName);
+                }
+            });
+            visitDepthFirst(*newClause, [&](const AstRecordInit& record) {
+                AstTypeIdentifier newName = binding.find(record.getType());
+                if (!newName.empty()) {
+                    const_cast<AstRecordInit&>(record).setType(newName);
+                }
+            });
             if (rel) {
```
Comment (mmcgr) [inline src/ComponentModel.cpp]: rel != nullptr

File: src/TypeConstraint.cpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    // update the type
+    currentSolution->setType(argument, newType);
+    assert(isSatisfied(currentSolution) && "union constraint resolution failed");
+}
+
+bool UnionConstraint::isSatisfied(const TypeSolver* currentSolution) const {
+    assert(currentSolution->hasType(argument) && "argument does not have an associated type");
+    const auto& bounds = getBounds();
+    for (const auto& bound : bounds) {
```
Comment (mmcgr) [inline src/TypeConstraint.cpp]: just iterate directly over getBounds()

File: tests/evaluation/magic_string_substr/magic_string_substr.dl [hunk 4]
```diff
@@ -26,4 +26,4 @@ C(substr("12",22,12)) :- Nullary().
 // TODO (lyndonhenry): The code following this comment is a hack to disable the `-e` option, which happens automatically when a program contains record types. The `substr` functor returns an empty string when called with an invalid index. As such, the invalid substring indices used on line 19 of this test will cause a tuple containing the empty string to occur in the relation of `C`, which is otherwise empty. Using the `-efile` and `-m*` options together causes the intermediate IDB file `C+_f.facts` to be written to `output-dir`, where this file contains a single blank line for the tuple with the empty string in `C`. When this same file is then read as intermediate EDB by successor strata, This is not a bug with either the `-efile` or `-m*` options, nor their use together, but rather something else. The issue is that when empty strings are derived as IDB they can be written as output, however empty strings cannot be read as input EDB.
```
Comment (mmcgr) [inline tests/evaluation/magic_string_substr/magic_string_substr.dl]: hmm. is this still true?

**Comment by azreika** [comment]
Note - not expected to pass the tests yet. More manageable pull requests will be made soon after with fixes.

File: src/TypeConstraint.cpp (hunk 3 above)
Comment (azreika) [reply src/TypeConstraint.cpp]: I've used bounds another time in that method, thought it was less clunky?

File: tests/evaluation/magic_string_substr/magic_string_substr.dl (hunk 4 above)
Comment (azreika) [reply tests/evaluation/magic_string_substr/magic_string_substr.dl]: Doesn't seem like it is, but this is independent of this PR I think. Should be changed in master.

**Pull request merged**